Monokai

epmoyer opened the pull request:
This change adds "monokai" and "monokai-256" to the built-in Themes. It faithfully reproduces all Monokai styling except for coloring optional argument names orange in function calls (implementing that would have required a two-pass parser to go backwards and change the token type of the argument name after encountering the "=" which follows it.)

This implementation adds several new palette items: "namespace", "operator", "argument", "builtin", "pseudo", "dunder", "exception", and "keyword2".  Their meaning is documented near the top of theme.py

All legacy Themes have been updated to assign colors to the new palette items such that the legacy themes appear identical before and after this change.  The addition of the new palette items makes it possible, if you like, to expand the colorization of the legacy themes in the future. I thought it was cleanest for me to keep this pull request focussed on the new theme and not muddy the waters by making any changes to the appearance of the legacy ones.

I have tested on Python 3.5 and 2.7, and raw/curses (though both appear identical on my platform).

**monokai-256 vs Sublime**
<img width="941" alt="monokai-256" src="https://cloud.githubusercontent.com/assets/136718/21089341/23d53790-bfec-11e6-985b-97d5b4691238.png">

**monokai vs Sublime**
<img width="921" alt="monokai" src="https://cloud.githubusercontent.com/assets/136718/21089343/29260ddc-bfec-11e6-8452-2153948ddd41.png">

**Legacy themes with Monokai build (left) and trunk build (right)**
<img width="1439" alt="classic_compare" src="https://cloud.githubusercontent.com/assets/136718/21089359/43dcf078-bfec-11e6-9580-dc2ab6205552.png">

<img width="1435" alt="vim_compare" src="https://cloud.githubusercontent.com/assets/136718/21089365/574e5c78-bfec-11e6-808c-21aabf266ac4.png">

<img width="1437" alt="dark_vim_compare" src="https://cloud.githubusercontent.com/assets/136718/21089370/6879634e-bfec-11e6-8a27-555e23be0401.png">

<img width="1439" alt="midnight_compare" src="https://cloud.githubusercontent.com/assets/136718/21089381/7b30e14c-bfec-11e6-91b8-c52611a17223.png">

<img width="1437" alt="solarized_compare" src="https://cloud.githubusercontent.com/assets/136718/21089396/926bfc52-bfec-11e6-8704-b9da78212ecb.png">

<img width="1436" alt="agr-256_compare" src="https://cloud.githubusercontent.com/assets/136718/21089412/b3d8873e-bfec-11e6-8c4d-8df3505044d5.png">

epmoyer commented:
Ack! Missed a dunder style in solarized and just noticed it in my screenshots.  Fixed now.

<img width="1438" alt="solarized_compare_2" src="https://cloud.githubusercontent.com/assets/136718/21089684/05d4b97a-bfef-11e6-8f0e-8c0391dafebd.png">

jeffwidman commented:
Fix #207

asmeurer commented on pudb/source_view.py:
```diff
@@ -222,3 +278,51 @@ def shipout_line():
         return result
+
+from enum import Enum
+class ParseState(Enum):
```
Enum is Python 3-only, but pudb needs to support Python 2 for the moment. I think this should still work just fine if you make it a regular class for how you are using it.

epmoyer replied:
Agreed!  I now have a 2.6 environment setup and have tested removing the enum with 2.6 & 3.5 (works).  Will push in a moment.

asmeurer commented on pudb/source_view.py:
```diff
@@ -222,3 +278,51 @@ def shipout_line():
+    found_function = 2
+    found_open_paren = 3
+
+class AgumentParser():
```
Subclass from `object` for Python 2 compatibility.

epmoyer replied:
Agreed

asmeurer commented:
Do the new tokens fallback to their counterparts automatically? If so, then the changes to the existing themes seems redundant. If not, then what will happen if someone has their own custom theme?

asmeurer commented:
I'm unclear if it's possible with the current code, but terminals can support italics (see https://web.archive.org/web/20161003094535/https://alexpearce.me/2014/05/italics-in-iterm2-vim-tmux/ for iTerm2, for instance).

epmoyer commented:
The new palette items do not automatically inherit from the old set, but I definitely see your point about the desirability of supporting existing custom themes without modification.  It seems achievable, since there is a direct inheritance mapping which would preserve the look of existing themes.  I'll take a crack at it tomorrow.  It's a fine idea, thanks!

And you're right, if the inheritance mapping was automatic then I could remove the new palette items from the legacy themes.

epmoyer commented:
The current release of urwid (1.3.1) does not contain support for italics.  Support for italics was committed on Apr 19, 2015 (and is part of the master branch now), but has not yet made it into a release (1.3.1 was released Nov 1, 2015, but did not include the italics commit).  I found that Apple’s terminal app does not support italics, but iTerm2 does (I was able to see italics there by printing the associated escape codes), so that’s a viable test environment whenever the next urwid release drops.

This prints italics in terminals that support it:
`print ('plain text \033[3m italic text \033[0m plain text’)`

And this will print in italics when the next version of urwid comes out:
```
import urwid

txt = urwid.Text(("mystyle", "Hello World"))
fill = urwid.Filler(txt, 'top')
palette = [('mystyle', 'default,italics', 'default'),]
loop = urwid.MainLoop(fill, palette)
loop.run()
```

asmeurer commented:
No need to necessarily add italics. Just wanted to point out that it was possible if you wanted to do it. I guess we should wait for urwid support.

epmoyer commented:
Added an inheritance mapping which assigns each new added style to the value of it's corresponding legacy style if a Theme does not assign that style explicitly.  With that in place, I regressed all the legacy themes to their original code (and they work properly).  The inheritance is applied after custom Themes are loaded, so custom Themes produce identical visual behavior without modification.

I also updated "monokai" to use a dark style for the side panels (like "monokai-256" does).

**monokai**
<img width="699" alt="monokai" src="https://cloud.githubusercontent.com/assets/136718/21166199/d8f193cc-c158-11e6-8acf-f1cc62d991ae.png">

**monokai-256**
<img width="698" alt="monokai-256" src="https://cloud.githubusercontent.com/assets/136718/21166205/e8d82166-c158-11e6-8fed-63a9483cd054.png">

**monokai build (left) vs master build (right) with an (ugly :) unmodified legacy custom theme**
<img width="1426" alt="compare_custom" src="https://cloud.githubusercontent.com/assets/136718/21166249/1619777e-c159-11e6-90de-ad930e1128f0.png">

**monokai build (left) vs master build (right) with legacy themes**
<img width="1423" alt="compare_classic" src="https://cloud.githubusercontent.com/assets/136718/21166283/53df2202-c159-11e6-9761-889a7acc34f4.png">

<img width="1422" alt="compare_vim" src="https://cloud.githubusercontent.com/assets/136718/21166288/5cd30784-c159-11e6-958d-8726baca60f4.png">

<img width="1421" alt="compare_dark_vim" src="https://cloud.githubusercontent.com/assets/136718/21166293/62e80c82-c159-11e6-8c12-57119fbbd537.png">

<img width="1427" alt="comare_midnight" src="https://cloud.githubusercontent.com/assets/136718/21166297/69b806fc-c159-11e6-86a5-f6b2102c19ac.png">

<img width="1427" alt="compare_solarized" src="https://cloud.githubusercontent.com/assets/136718/21166298/6e751c16-c159-11e6-8baa-cbd65e9dce07.png">

<img width="1425" alt="compare_agr-256" src="https://cloud.githubusercontent.com/assets/136718/21166303/767de1cc-c159-11e6-9f98-dd9ba026df00.png">

inducer commented on pudb/source_view.py:
```diff
@@ -151,14 +151,34 @@ def format_source(debugger_ui, lines, breakpoints):
 
         result = []
-
+        argument_parser = AgumentParser(t)
```
`Agument` -> `Argument`

inducer commented on pudb/source_view.py:
```diff
@@ -222,3 +278,50 @@ def shipout_line():
+
+    This parser implements a state machine which accepts
+    Pygments tokens, delivered sequentially from the beginning
+    of a source file to it's end.
```
it's -> its

inducer commented:
Looks great. Thanks for contributing this!

epmoyer commented:
My pleasure! Thanks Andreas for making such an awesome tool and for being so responsive.  And thanks Aaron for the excellent comments & responsiveness; the patch ended up much better for your suggestions.

Cheers everyone!

jeffwidman commented:
Thanks @epmoyer really high quality PR that I'm looking forward to using
